test/testrelative.c

```diff
@@ -21,39 +22,72 @@
 
 static SDLTest_CommonState *state;
 static int i, done;
-static float mouseX, mouseY;
 static SDL_FRect rect;
 static SDL_Event event;
+static SDL_bool warp;
 
 static void DrawRects(SDL_Renderer *renderer)
 {
     SDL_SetRenderDrawColor(renderer, 255, 0, 0, 255);
-    rect.x = mouseX;
-    rect.y = mouseY;
     SDL_RenderFillRect(renderer, &rect);
+
+    SDL_SetRenderDrawColor(renderer, 255, 255, 255, 255);
+    if (SDL_GetRelativeMouseMode()) {
+        SDLTest_DrawString(renderer, 0.f, 0.f, "Relative Mode: Enabled");
+    } else {
+        SDLTest_DrawString(renderer, 0.f, 0.f, "Relative Mode: Disabled");
+    }
 }
 
 static void loop(void)
 {
+    SDL_Window *window = SDL_GetMouseFocus();
+
     /* Check for events */
     while (SDL_PollEvent(&event)) {
         SDLTest_CommonEvent(state, &event, &done);
         switch (event.type) {
+        case SDL_EVENT_KEY_DOWN:
+            if (event.key.key == SDLK_C) {
+                /* If warp emulation is active, showing the cursor should turn
+                 * relative mode off, and it should re-activate after a warp
+                 * when hidden again.
+                 */
+                if (SDL_CursorVisible()) {
+                    SDL_HideCursor();
+                } else {
+                    SDL_ShowCursor();
+                }
+            }
+            break;
         case SDL_EVENT_MOUSE_MOTION:
         {
-            mouseX += event.motion.xrel;
-            mouseY += event.motion.yrel;
+            rect.x += event.motion.xrel;
+            rect.y += event.motion.yrel;
         } break;
         default:
             break;
         }
     }
+
     for (i = 0; i < state->num_windows; ++i) {
         SDL_Rect viewport;
         SDL_Renderer *renderer = state->renderers[i];
         if (state->windows[i] == NULL) {
             continue;
         }
+
+        if (warp && (SDL_GetWindowFlags(state->windows[i]) & SDL_WINDOW_INPUT_FOCUS)) {
```

Comment:
Can you move this to the motion event handler for a better representation of what games do? Maybe also note in the comment below that this isn't a recommended way of doing relative mouse motion? :)

Reply:
Done!